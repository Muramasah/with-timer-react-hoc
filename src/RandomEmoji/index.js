// THIRD PARTY DEPENDENCIES
import React, { useEffect, useState } from 'react';
import axios from 'axios';

// INTERNAL DEPENDENCIES
import withTimer from '../withTimer';

// CONSTANTS
const API_URL = 'https://api.github.com/emojis';
const DEFAULT_INTERVAL = 1000;
const STYLES = {
  CONTAINER: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    height: '100vh'
  },
  EMOJI: {
    width: '64px'
  }
};

// SELECTORS
const selectRandomEmoji = (emojis) => {
  const names = Object.keys(emojis);

  return emojis[names[(names.length * Math.random()) << 0]];
};

// COMPONENT
const RandomEmoji = (props) => {
  // PROPERTIES
  const { clearTimer, setTimer } = props;

  // STATE
  const [emojis, setEmojis] = useState({});
  const [emoji, setEmoji] = useState('');

  // PRIVATE METHODS
  const mutateEmoji = () => setEmoji(selectRandomEmoji(emojis));
  const mutateEmojis = serverEmojis => {
    setEmojis(serverEmojis);
    mutateEmoji();
  };

  // LIFE CYCLE
  const fetchEmojisEffectHandler = () => {
    // EFFECT
    axios.get(API_URL)
      .then(response => response.json())
      .then(mutateEmojis);

    // EFFECTS CLEANER
    /** We do not needmutateEmojismutateEmojis this to return any effect cleaner */
  }
  const setTimerEffectHandler = () => {
    // EFFECT
    setTimer(mutateEmoji, DEFAULT_INTERVAL); /** Here we set the timer */

    // EFFECTS CLEANER
    return clearTimer /** Here we return the cleaner function. */
  }

  useEffect(fetchEmojisEffectHandler);
  useEffect(setTimerEffectHandler);

  // VISUAL NODES
  return (
    <div style={STYLES.CONTAINER}>
      <img src={emoji} style={STYLES.EMOJI} />
    </div>
  );
};

const RandomEmojiWithTimer = withTimer(RandomEmoji);

// EXPORT
export { RandomEmoji, RandomEmojiWithTimer };
export default RandomEmojiWithTimer;
