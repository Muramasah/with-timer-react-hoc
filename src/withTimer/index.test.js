// THIRD PARTY DEPENDENCIES
import React from 'react';
import renderer from 'react-test-renderer';

// INTERNAL DEPENDENCIES
import RandomEmoji from '../RandomEmoji';

test('An Emoji is displayed', () => {
  const component = renderer.create(<RandomEmoji />);
  let tree = component.toJSON();
  
  expect(tree).toMatchSnapshot();
});