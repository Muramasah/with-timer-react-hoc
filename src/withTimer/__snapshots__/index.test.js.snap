// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An Emoji is displayed 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100vh",
      "justifyContent": "center",
    }
  }
>
  <img
    src=""
    width="64px"
  />
</div>
`;
